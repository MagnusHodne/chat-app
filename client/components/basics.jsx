import React from "react";
import { Link, useNavigate } from "react-router-dom";

export function FAIcon({ icon, className }) {
  return <i className={`icon ${icon} ${className}`}></i>;
}

export function Button({ title, icon, to, onClick, className }) {
  if (to) {
    const navigate = useNavigate();
    onClick = () => {
      navigate(to);
    };
  }
  const button = () => (
    <button
      className={`button ${className !== undefined ? className : ""}`}
      onClick={onClick}
    >
      {icon && <FAIcon icon={icon} />}
      {title}
    </button>
  );
  return button();
}
