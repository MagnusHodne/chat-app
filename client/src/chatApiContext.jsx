import React from "react";
import {
  deleteJSON,
  fetchJSON,
  postJSONNoReturn,
  putJSON,
} from "./lib/fetchUtils";
import { randomString } from "./lib/randomString";
import { sha256 } from "./lib/sha256";

export const ChatApiContext = React.createContext({
  /**
   * Fetches login info from the server (body contains both configs for the different providers, as
   * well as the user info (if any))
   */
  async fetchConfig() {
    return await fetchJSON("/api/login/config");
  },
  /**
   * Sends the access_token from a provider to the server (so that it can be signed)
   * @param provider the provider that has been used
   * @param login the login info, which must contain an access_token
   */
  async registerLogin(provider, login) {
    return await postJSONNoReturn(`/api/login/${provider}`, login);
  },
  /**
   * The actual method that handles login. Redirects to the provider endpoint
   * @param provider
   * @param config The config object, obtained from the server
   */
  async handleLogin(provider, config) {
    const {
      authorization_endpoint,
      response_type,
      scope,
      client_id,
      code_challenge_method,
    } = config[provider];

    //Code challenge state?
    const state = randomString(50);
    window.sessionStorage.setItem("expected_state", state);

    //The url parameters we send to the provider
    const params = {
      response_type,
      response_mode: "fragment",
      client_id,
      state,
      scope,
      redirect_uri: `${window.location.origin}/login/${provider}/callback`,
    };

    //Code challenge stuff, if applicable
    if (code_challenge_method) {
      const code_verifier = randomString(50);
      window.sessionStorage.setItem("code_verifier", code_verifier);
      params.code_challenge_method = code_challenge_method;
      params.code_challenge = await sha256(code_verifier);
    }

    //Redirecting the user to the authorization endpoint
    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(params);
  },
  getCallbackParameters() {
    return Object.fromEntries(
      new URLSearchParams(window.location.hash.substring(1))
    );
  },
  getExpectedState() {
    return window.sessionStorage.getItem("expected_state");
  },

  /**
   * Makes the server delete our cookies (effectively logging us out)
   */
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
  verifyUser(user) {
    if (!user) return false;
    return Object.entries(user).length !== 0;
  },
  /*==== DATABASE OPERATIONS =====*/
  async fetchChatLog() {
    return await fetchJSON("/api/chat");
  },
  async fetchUserInfo({ sub }) {
    return await fetchJSON(`/api/user?sub=${sub}`);
  },
  async updateUserBio({ sub, bio }) {
    return await putJSON("/api/user", { sub, bio });
  },
  async deleteMessage({ _id }) {
    return await deleteJSON("/api/chat", { _id });
  },
});
