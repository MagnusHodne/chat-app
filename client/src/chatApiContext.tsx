import React from "react";
import { deleteJSON, fetchJSON, putJSON } from "./lib/fetchUtils";

export const ChatApiContext = React.createContext({
  /*==== DATABASE OPERATIONS =====*/
  async fetchChatLog() {
    return await fetchJSON("/api/chat");
  },
  async fetchUserInfo({ sub }: { sub: string }) {
    return await fetchJSON(`/api/user?sub=${sub}`);
  },
  async updateUserBio({ sub, bio }: { sub: string; bio: string }) {
    return await putJSON("/api/user", { sub, bio });
  },
  async deleteMessage({ _id }: { _id: any }) {
    return await deleteJSON("/api/chat", { _id });
  },
});
