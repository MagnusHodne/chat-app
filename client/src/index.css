@import url("https://fonts.googleapis.com/css2?family=Nunito:ital,wght@0,200;0,300;0,400;0,500;0,600;0,700;0,800;0,900;1,200;1,300;1,400;1,500;1,600;1,700;1,800;1,900&family=Open+Sans:ital,wght@0,300;0,400;0,500;0,600;0,700;0,800;1,300;1,400;1,500;1,600;1,700;1,800&display=swap");
@import "../node_modules/@fortawesome/fontawesome-free/css/all.css";
@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .scrollbar::-webkit-scrollbar {
    @apply w-2.5;
  }
  .scrollbar::-webkit-scrollbar-track {
    @apply bg-thischord-800;
    @apply rounded-full;
    @apply my-2;
  }

  .scrollbar::-webkit-scrollbar-thumb {
    @apply bg-thischord-900;
    @apply rounded-full;
  }

  .scrollbar::-webkit-scrollbar-thumb:hover {
    background: #c0a0b9;
  }
}

html {
  @apply text-white;
}
