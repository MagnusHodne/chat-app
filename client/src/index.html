<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Chat application PGR6301</title>
    <link rel="stylesheet" href="index.css">
</head>
<body class="bg-thischord-900">
<div id="app"></div>
<script src="index.jsx" type="module"></script>
</body>
</html>