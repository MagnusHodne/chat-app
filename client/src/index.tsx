import ReactDOM from "react-dom/client";
import { Application } from "./application";
import { Auth0Config } from "../auth0.config";

const root = ReactDOM.createRoot(
  document.getElementById("root") as HTMLElement
);
root.render(
  <Auth0Config>
    <Application />
  </Auth0Config>
);
