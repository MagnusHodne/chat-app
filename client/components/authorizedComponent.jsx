import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoadingComponent } from "./feedbackComponents";
import { ChatApiContext } from "../chatApiContext";

/**
 * Simple wrapper for components that require the user to be authorized
 * @param user the user object used for checking verification
 * @param {boolean} writeAccess Whether this component needs write access (default is false)
 * @param successComponent
 * @param failComponent
 * @param verificationFunction what function to use for verifying user. By default it uses the one specified in apiMethods
 */
export function AuthorizedComponent({
  user,
  writeAccess = false,
  successComponent = <div>Success</div>,
  failComponent = <div>Fail</div>,
}) {
  const navigate = useNavigate();
  const [verified, setVerified] = useState(false);
  const { verifyUser } = useContext(ChatApiContext);
  useEffect(() => {
    if (!verifyUser(user)) {
      navigate("/startlogin");
    } else {
      setVerified(true);
    }
  }, []);
  if (!verified)
    return <LoadingComponent message={"Redirecting to login..."} />;
  return successComponent;
}
