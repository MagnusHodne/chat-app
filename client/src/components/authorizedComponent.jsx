import React, { useContext, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { LoadingComponent } from "./feedbackComponents";
import { ChatApiContext } from "../chatApiContext";
import { UserContext } from "../userContext";

/**
 * Simple wrapper for components that require the user to be authorized
 * @param user the user object used for checking verification
 * @param {boolean} writeAccess Whether this component needs write access (default is false)
 * @param successComponent
 * @param failComponent
 */
export function AuthorizedComponent({
  writeAccess = false,
  successComponent = <div>Success</div>,
  failComponent = <div>Fail</div>,
}) {
  const navigate = useNavigate();
  const [verified, setVerified] = useState(false);
  const { user } = useContext(UserContext);
  const { verifyUser } = useContext(ChatApiContext);
  useEffect(() => {
    if (!verifyUser(user)) {
      navigate("/startlogin");
    } else {
      setVerified(true);
    }
  }, []);
  if (!verified)
    return <LoadingComponent message={"Redirecting to login..."} />;
  return successComponent;
}
