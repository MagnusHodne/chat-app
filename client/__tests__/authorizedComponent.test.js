import React from "react";
import ReactDOM from "react-dom";
import { AuthorizedComponent } from "../src/components/authorizedComponent";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { act } from "react-dom/test-utils";
import { ChatApiContext } from "../src/chatApiContext";

let element;
beforeEach(() => {
  element = document.createElement("div");
});

describe("AuthorizedComponent", () => {
  it("uses an authorization function", async () => {
    const verifyUser = jest.fn(() => false);

    await act(async () => {
      ReactDOM.render(
        <MemoryRouter>
          <ChatApiContext.Provider value={{ verifyUser }}>
            <AuthorizedComponent user={{ name: "test-user" }} />
          </ChatApiContext.Provider>
        </MemoryRouter>,
        element
      );
    });

    expect(verifyUser).toHaveBeenCalled();
    expect(element.innerHTML).toMatchSnapshot();
    expect(element.querySelector("p").innerHTML).toContain("Redirect");
  });

  it("returns the content on successful authorization", async () => {
    const verifyUser = jest.fn(() => true);
    function TestElement() {
      return <div className={".test-component"}>Test success component</div>;
    }
    await act(async () => {
      ReactDOM.render(
        <BrowserRouter>
          <ChatApiContext.Provider value={{ verifyUser }}>
            <AuthorizedComponent
              successComponent={<TestElement />}
              user={{ name: "test-user" }}
            />
          </ChatApiContext.Provider>
        </BrowserRouter>,
        element
      );
    });

    expect(verifyUser).toHaveBeenCalled();
    expect(element.innerHTML).toMatchSnapshot();
    expect(element.querySelector("div").innerHTML).toContain(
      "Test success component"
    );
  });
});
