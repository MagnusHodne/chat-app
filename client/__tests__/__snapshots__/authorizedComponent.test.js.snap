// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizedComponent returns the content on successful authorization 1`] = `"<div class=\\".test-component\\">Test success component</div>"`;

exports[`AuthorizedComponent uses an authorization function 1`] = `"<div class=\\"loading-component flex h-full items-center justify-center\\"><div class=\\"flex flex-row items-center gap-2\\"><span class=\\"fa-solid fa-spinner fa-spin-pulse\\"></span><p>Redirecting to login...</p></div></div>"`;
